### PR TITLE
Vectara

### PR DESCRIPTION
**Pull request by ofermend**
# Vectara Integration

This PR provides integration with Vectara. Implemented here are:
* langchain/vectorstore/vectara.py
* tests/integration_tests/vectorstores/test_vectara.py
* langchain/retrievers/vectara_retriever.py
And two IPYNB notebooks to do more testing:
* docs/modules/chains/index_examples/vectara_text_generation.ipynb
* docs/modules/indexes/vectorstores/examples/vectara.ipynb

### CODE REVIEW

File: langchain/vectorstores/vectara.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import json
+import logging
+from hashlib import md5
+import requests
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.vectorstores.base import VectorStore, VectorStoreRetriever
+
+"""Implementation of Vector Store using Vectara (https://vectara.com)"""
```
Comment (dev2049) [inline langchain/vectorstores/vectara.py:14]: docstring a bit out of place (if here should be comment)

File: langchain/retrievers/vectara_retriever.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+"""Milvus Retriever"""
+from typing import Optional, List
+
+from langchain.schema import BaseRetriever, Document
+from langchain.vectorstores.vectara import Vectara
+
+class VectaraRetriever(BaseRetriever):
```
Comment (dev2049) [inline langchain/retrievers/vectara_retriever.py:7]: we probably want to make Vectara.as_retriever() return this - so i would move it into that file (similar to how we do with redis)

**Review by dev2049 — COMMENTED** [review]
the two notebooks in chains/index_examples are a bit out place - we can't really have that for all integrations. if you want to keep around, i would move into the integrations directory and link to from the integrations page

File: langchain/retrievers/vectara_retriever.py (hunk 2 above)
Comment (ofermend) [reply langchain/retrievers/vectara_retriever.py:7]: so just move VectaraRetriever to vectara.py and get rid of vectara_retriever.py altogether?

Comment (dev2049) [reply langchain/retrievers/vectara_retriever.py:7]: yep!

File: langchain/vectorstores/vectara.py (hunk 1 above)
Comment (ofermend) [reply langchain/vectorstores/vectara.py]: > the two notebooks in chains/index_examples are a bit out place - we can't really have that for all integrations. if you want to keep around, i would move into the integrations directory and link to from the integrations page

Moved the notebooks to integrations. Thanks!

Comment (ofermend) [reply langchain/vectorstores/vectara.py]: > docstring a bit out of place (if here should be comment)

Yep. fixed!

File: langchain/retrievers/vectara_retriever.py (hunk 2 above)
Comment (ofermend) [reply langchain/retrievers/vectara_retriever.py]: Perfect. Done!

File: langchain/vectorstores/vectara.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            vectorstore = Vectara(customer_id, corpus_id, api_key)
+    """
+
+    def __init__(
+        self, vectara_customer_id: int, vectara_corpus_id: int, vectara_api_key: str
+    ):
+        """Initialize with Vectara API."""
+        self._vectara_customer_id = vectara_customer_id
+        self._vectara_corpus_id = vectara_corpus_id
+        self._vectara_api_key = vectara_api_key
```
Comment (dev2049) [inline langchain/vectorstores/vectara.py:31]: we usually make it possible to set credentials (like api key and other id's) via environment variables. if we wanted to support that here could d something like
```python
def __init__(vectara_api_key: Optional[str]=None, ...):
  self.vectara_api_key = vectara_api_key or get_from_env("vectara_api_key", "VECTARA_API_KEY")
  ...
```

File: tests/integration_tests/vectorstores/test_vectara.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        metadatas=None,
+        customer_id=customer_id,
+        corpus_id=corpus_id,
+        api_key=api_key,
+    )
+
+    new_texts = ["foobar", "foobaz"]
+    docsearch.add_documents([Document(page_content=content) for content in new_texts])
+    output = docsearch.similarity_search("foobar", k=1)
+    assert output == [Document(page_content="foobar")] or output == [
```
Comment (dev2049) [inline tests/integration_tests/vectorstores/test_vectara.py:26]: is this not deterministic?

**Review by dev2049 — COMMENTED** [review]
two quick comments

**Review by dev2049 — COMMENTED** [review]
two quick comments

File: langchain/vectorstores/vectara.py (hunk 3 above)
Comment (ofermend) [reply langchain/vectorstores/vectara.py]: Good suggestion - updated!

File: tests/integration_tests/vectorstores/test_vectara.py (hunk 4 above)
Comment (ofermend) [reply tests/integration_tests/vectorstores/test_vectara.py]: Good point. I updated the test to not only be deterministic but also more comprehensive and test metadatas.

**Pull request merged**